dashboards/storage/storage-io-namespace.libsonnet

```diff
@@ -0,0 +1,356 @@
+          skipUrlSync: false,
+        };
+
+      local resolutionTemplate =
+        template.new(
+          name='resolution',
+          datasource='$datasource',
+          query='1m,5m,1h',
+          current='5m',
+          hide='',
+          refresh=2,
+          includeAll=false,
+          sort=1
+        ) + {
+          auto: false,
+          auto_count: 30,
+          auto_min: '10s',
+          skipUrlSync: false,
+          type: 'interval',
+          options: [
+            {
+              selected: false,
+              text: '1m',
+              value: '1m',
+            },
+            {
+              selected: true,
+              text: '5m',
+              value: '5m',
+            },
+            {
+              selected: false,
+              text: '1h',
+              value: '1h',
+            },
+          ],
+        };
+
+      local intervalTemplate =
+        template.new(
+          name='interval',
+          datasource='$datasource',
+          query='4h',
+          current='5m',
+          hide=2,
+          refresh=2,
+          includeAll=false,
+          sort=1
+        ) + {
+          auto: false,
+          auto_count: 30,
+          auto_min: '10s',
+          skipUrlSync: false,
+          type: 'interval',
+          options: [
+            {
+              selected: true,
+              text: '4h',
+              value: '4h',
+            },
+          ],
+        };
+
+      //#####  IOPS Row ######
+
+      local iopsRow =
+        row.new(
+          title='IOPS'
+        );
+
+      //#####  Throughput Row ######
+
+      local throughputRow =
+        row.new(
+          title='Throughput'
+        );
+
+      dashboard.new(
+        title='%(dashboardNamePrefix)sStorage / Namespace (Pods IO)' % $._config.grafanaK8s,
+        tags=($._config.grafanaK8s.dashboardTags),
+        editable=true,
+        schemaVersion=18,
+        refresh=($._config.grafanaK8s.refresh),
+        time_from='now-1h',
+        time_to='now',
+      )
+      .addTemplate(
+        {
+          current: {
+            text: 'default',
+            value: 'default',
+          },
+          hide: 0,
+          label: null,
+          name: 'datasource',
+          options: [],
+          query: 'prometheus',
+          refresh: 1,
+          regex: '',
+          type: 'datasource',
+        },
+      )
+      .addTemplate(clusterTemplate)
+      .addTemplate(namespaceTemplate)
+      .addTemplate(resolutionTemplate)
+      .addTemplate(intervalTemplate)
+      .addAnnotation(annotation.default)
+      .addPanel(iopsRow, gridPos={ h: 1, w: 24, x: 0, y: 0 })
+      .addPanel(
+        newGraphPanel(
+          graphTitle='Reads',
+          graphQuery='sum by(namespace, pod) (irate(container_fs_reads_total{container!="", %(clusterLabel)s="$cluster",namespace=~"$namespace"}[$interval:$resolution]))' % $._config,
```

Comment:
Why `irate` and not `rate` and why doing this with subqueries and not just interval?

Same question for all expressions used in this dashboard.

Reply:
@paulfantom There is no great reason for using irate, it just might capture spikes better. But if there is a good reason not to use it or the rate might give a better outcome, I would be happy to change it.

Regarding the resolution, I just took inspiration from the network dashboard. But now that you have pointed, I think setting the interval to a constant [1m] and removing the resolution might make more sense. Am I thinking on the right lines from your perspective?
Eg: irate(container_fs_writes_bytes_total[1m]) and may be saving these as recording rules ?

Reply:
We mostly set intervals to 5m to allow longer scrape intervals. With a 1m interval we need scrapes to happen at least every 30s for irate and 15s interval for rate.

I was thinking more in direction of `rate(container_fs_reads_bytes_total[5m])` but if the intention is catching spikes then `irate` might be better.

As for the recording rule, yes it might be good to have it considering the number of similar queries.

Reply:
All the metrics for these dashboards are from Cadvisor for which the default scrape is 30s, so I think we can use irate and [1m]. Thank you for pointing that out.
I will create a recording rule on the container level as I would be able to use that in a dashboard where a user wants to look under a pod and information on a container level.

I will make the required changes and will request another review. Thank you for all the suggestions.

Reply:
Yes, the default is 30s, but this is customizable and users may adjust it to a bit longer periods while still having meaningful data. Since we are depending on this interval, we need to either make it customizable or automatically adjustable (there is [__rate_interval variable in grafana 7.2](https://grafana.com/blog/2020/09/28/new-in-grafana-7.2-__rate_interval-for-prometheus-rate-queries-that-just-work/) and we can use it via `$._config.grafanaIntervalVar`) or make it longer so it won't break in scenarios where scrape_interval is higher.

In kubernetes-mixins we mostly choose `$._config.grafanaIntervalVar` or longer intervals.

Reply:
@paulfantom I did some more experiments, rate and [5m], interval gave much better values.